src/qtwebengine/dataprovider.cpp

```diff
@@ -1,20 +1,66 @@
+/*
+ *  Kchmviewer - a CHM and EPUB file viewer with broad language support
+ *  Copyright (C) 2004-2016 George Yunaev, gyunaev@ulduzsoft.com
+ *  Copyright (C) 2021 Nick Egorrov, nicegorov@yandex.ru
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
 #include <QBuffer>
+#include <QWebEngineUrlScheme>
 #include <QWebEngineUrlRequestJob>
 
-#include "mainwindow.h"
-#include "dataprovider_qwebengine.h"
+#include "../mainwindow.h"
+#include "dataprovider.h"
+#include "ebook_chm.h"
+#include "ebook_epub.h"
+
+#define PRINT_DEBUG ( defined PRINT_DEBUG_ALL || defined PRINT_DEBUG_WEBENGINE || defined PRINT_DEBUG_WEBENGINEDATAPROVIDER )
+
+static struct RegistrationHelper
+{
+    RegistrationHelper()
+    {
+        QWebEngineUrlScheme scheme( DataProvider::URL_SCHEME_EPUB );
+        scheme.setSyntax( QWebEngineUrlScheme::Syntax::HostAndPort );
+        scheme.setDefaultPort( 443 );
+        scheme.setFlags( QWebEngineUrlScheme::SecureScheme );
+        QWebEngineUrlScheme::registerScheme( scheme );
+
+        scheme.setName( DataProvider::URL_SCHEME_CHM );
+        QWebEngineUrlScheme::registerScheme( scheme );
+    }
+} helper;
 
-DataProvider_QWebEngine::DataProvider_QWebEngine( QObject *parent )
+const char * DataProvider::URL_SCHEME_CHM    = EBook_CHM::URL_SCHEME_CHM;
+const char * DataProvider::URL_SCHEME_EPUB   = EBook_EPUB::URL_SCHEME_EPUB;
+
+DataProvider::DataProvider( QObject *parent )
     : QWebEngineUrlSchemeHandler( parent )
 {
 }
 
-void DataProvider_QWebEngine::requestStarted( QWebEngineUrlRequestJob *request )
+void DataProvider::requestStarted( QWebEngineUrlRequestJob *request )
 {
     QUrl url = request->requestUrl();
-    bool htmlfile = url.path().endsWith( ".html" ) || url.path().endsWith( ".htm" ) || url.path().endsWith( ".xhtml" );
-
-    qDebug("requestStarted %s", qPrintable(url.toString()) );
+    QByteArray headerAccept = request->requestHeaders().value( "Accept" );
+#if PRINT_DEBUG
+    qDebug() << "[DEBUG] DataProvider::requestStarted";
+    qDebug() << "  url = " << url.toString();
+    qDebug() << "  Header \"Accept\" = " << headerAccept;
+#endif
+    bool htmlfile = headerAccept.contains( "text/html" );
```

Comment:
Can you please point out to the documentation which states this works as expected? For me it did not work.

Reply:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept
https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types#texthtml

The [QWebEngineUrlRequestJob::requestHeaders()](https://doc.qt.io/qt-5/qwebengineurlrequestjob.html#requestHeaders) function was introduced in Qt 5.13, while I was aiming for the 5.9 minimal version. I guess I'll have to look for another solution.

Reply:
This is not what I asked. I asked if you're sure whether Accept: header is correctly set in this scenario in every single test case.

Reply:
I'm also puzzled, same as with many others, **why** this change was made? Lots of them seem to be made without any particular purpose. It is unclear what issue, if any, they fix.